### PR TITLE
PTR-27_Setter_Based_Dependency_Injection

### PR DESCRIPTION
**Pull request by mitssev**

### CODE REVIEW

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java [hunk 1]
```diff
@@ -27,6 +27,10 @@ public AnnotationBeanFactory(BeanDefinitionRegistry registry, List<BeanPostProce
 
     @Override
     public <T> T getBean(Class<T> beanType) throws NoSuchBeanException, NoUniqueBeanException {
+        Set<BeanDefinition> beanDefinitions = registry.getBeanDefinitionByType(beanType);
+        if (beanDefinitions.size() > 1 ) {
+            ensureBeanDefinitionsCreated(beanDefinitions);
```
Comment (petrosteam) [inline src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:32]: Looks like, here we create beans and method createBean(Class<?> clazz) is never used.

Comment (mykolad4) [inline src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:32]: Looks like, here we create beans and method createBean(Class<?> clazz) is never used.

Comment (mykolad4) [reply src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:32]: When context is initialised, all our beans are instantiated. So this code can be useful only during application startup. Am I right?

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java [hunk 2]
```diff
@@ -59,14 +73,14 @@ private <T> T createBean(Class<T> beanType) {
             } else {
                 obj = beanType.getConstructor().newInstance();
             }
+            rootContextMap.put(beanDefinition.getName(), obj);
```
Comment (mykolad4) [inline src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:76]: We have to remove prototype bean from rootContextMap after postProcessors. Or we can use another map for storing temporary beans, and fully initialised objects would be stored into context after postProcessors.

File: src/main/java/com/petros/bring/context/AnnotationConfigApplicationContext.java [hunk 3]
```diff
@@ -20,20 +19,15 @@ public AnnotationConfigApplicationContext(BeanDefinitionRegistry registry, List<
     }
 
     private void registerBean(BeanDefinition beanDefinition) {
-        validate(beanDefinition);
-        getBean(registry.getBeanTypeByName(beanDefinition.getName()));
+        if (needsRegistration(beanDefinition)) {
+            getBean(registry.getBeanTypeByName(beanDefinition.getName()));
```
Comment (mykolad4) [inline src/main/java/com/petros/bring/context/AnnotationConfigApplicationContext.java:23]: Good idea, but still it is possible that we have to create prototype objects during startup if singleton bean depends on it.

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java (hunk 1 above)
Comment (mitssev) [reply src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:32]: this snipped I added for the case when many (>1) beans are found by incoming type (interface) to address case when @Primary bean metadata exists but in context map is yet created.

File: src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java (hunk 2 above)
Comment (mitssev) [reply src/main/java/com/petros/bring/bean/factory/AnnotationBeanFactory.java:76]: added new instance creation for prototype case

**Review by mykolad4 — APPROVED** [review]

**Review by petrosteam — APPROVED** [review]

**Pull request merged**